plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3Service.java

```diff
@@ -189,12 +203,32 @@ static ClientConfiguration buildConfiguration(S3ClientSettings clientSettings) {
         clientConfiguration.setResponseMetadataCacheSize(0);
         clientConfiguration.setProtocol(clientSettings.protocol);
 
-        if (Strings.hasText(clientSettings.proxyHost)) {
-            // TODO: remove this leniency, these settings should exist together and be validated
-            clientConfiguration.setProxyHost(clientSettings.proxyHost);
-            clientConfiguration.setProxyPort(clientSettings.proxyPort);
-            clientConfiguration.setProxyUsername(clientSettings.proxyUsername);
-            clientConfiguration.setProxyPassword(clientSettings.proxyPassword);
+        if (clientSettings.proxySettings != ProxySettings.NO_PROXY_SETTINGS) {
+            if (clientSettings.proxySettings.getType() == ProxySettings.ProxyType.SOCKS) {
+                SocketAccess.doPrivilegedVoid(() -> {
+                    if (clientSettings.proxySettings.isAuthenticated()) {
+                        Authenticator.setDefault(new Authenticator() {
+                            @Override
+                            protected PasswordAuthentication getPasswordAuthentication() {
+                                return new PasswordAuthentication(
+                                    clientSettings.proxySettings.getUsername(),
+                                    clientSettings.proxySettings.getPassword().toCharArray()
+                                );
+                            }
+                        });
+                    }
+                    clientConfiguration.getApacheHttpClientConfig()
+                        .setSslSocketFactory(createSocksSslConnectionSocketFactory(clientSettings.proxySettings.getAddress()));
```

Comment:
Discussed that with @willyborankin offline: the `SslSocketFactory` is only used for HTTPS connections, otherwise `ApacheHttpClient` will use `SdkPlainSocketFactory` and the trick with opening `new Socket(new Proxy(Proxy.Type.SOCKS, address))` would not work. It is probably make sense to thrown an exception in this case (non-HTTPS endpoints)